feat: add liqo instance info page

Deblazz opened the pull request:
This PR will track my advancements regarding the Liqo status page.

Following a call with FrancescoC, we identified a couple of points to address:

- The data should be accessed using the _/v1/clusters_ endpoint and thus I should differentiate whether I am in a consumer or in a producer cluster.
- I can use the various APIs offered by liqoctl (as is currently done [here](https://github.com/Deblazz/liqo-dashboard/blob/1d331ba240e9fd0dcbf8a15770285602d3387e13/backend/pkg/server/handlers/status.go#L25) to obtain the data.

Based on these directions, I will continue the development of the status page.
Feedback and suggestions are very welcome.

Thanks!

Deblazz committed:
feat: add liqo instance info page

frisso commented:
I really like this.
It would be nice to have also the detailed view of the other clusters (if I remember correctly, "liqoctl info peers" shows a lot of information as well.
Thanks!

Deblazz commented:
Thanks for your feedback.
I'm currently working on adding those infos as well.
I will follow up asap.

Deblazz commented:
In the end, what do we think is the best approach? I could either add more details to the table, or include (for each peering) a button linking to the page suggested by @claudiolor.
I’ll wait for your thoughts.
Thanks!

claudiolor commented:
Hi @Deblazz,
Regarding the PR, overall the UI looks good to me. However, I think we could slightly reorganize the information displayed in the boxes at the top of the page. We can still keep three sections, structured as follows:

- **Liqo instance**
  - Cluster ID
  - Version
  - Liqo health (missing at the moment_
- Local cluster info:
  - K8s api server addr
  - Cluster labels
- Network details  (considering that you might have more then one network in these fields)
   - Pod CIDR
   - Service CIDR
   - External CIDR
   
As for the way we are exposing data in the backend, I’m not fully convinced by the current solution. I would prefer adding a dedicated route to expose the data about the current Liqo installation, rather than including the local installation details in the /clusters route, mainly because of the following reasons:

- We are representing two different concepts: what you call "local" refers to the current Liqo instance, while "remote" refers to peerings with external clusters. It’s difficult to design a schema that fits both, and currently some values don’t make sense in the “Local cluster” context (e.g., the status of network, authentication, and offloading modules).
- You will always have just one "local cluster".
- Every time you call the `/clusters` endpoint, you need to correctly set the `clusterType` query parameter, which is error-prone.

All in all, these are the changes I suggest:
- **Create an endpoint specific for the Liqo instance info** like `/info` where we can return the data in the same format `liqoctl info --verbose -o json` already provides (I already talked with @cheina97 and he agrees with this approach)
- Reorganize the data in the boxes at the top of the page as above
- Add, for each entry in the table summarizing active peerings, a link to the detailed peering page.
- Optionally, we could also separate active peerings from inactive ones (the ones where all modules are disabled) by displaying them in a different table. This is something that we should do also in the liqoctl command

frisso commented:
This is so cool, @Deblazz !
One more question, though: how are you planning to show IP addresses in cases there are more ranges? (not 100% sure, but it seems to be that we may have more than one range per each type... asking confirmation to @claudiolor )

Deblazz commented:
Hi,
currently this issue is not managed.
I'm not really sure as to how I should do it considering that, when it comes to network values, all of the data received from the JSON is a string, not a list of elements/strings (as is the case for the peers). For reference the following is the output of the _liqoctl info --verbose -o json_ on my setup.
For network parameters, I don't have lists:
`  "network": {
    "podCIDR": "10.244.0.0/16",
    "serviceCIDR": "10.96.0.0/16",
    "externalCIDR": "10.70.0.0/16",
    "internalCIDR": "10.80.0.0/16"
  }`
  For peers:
  `
  "peerings": {
    "peers": [
      {
        "clusterID": "liqo-peer-1",
        "role": "Provider",
        "networkingStatus": "Healthy",
        "authenticationStatus": "Healthy",
        "offloadingStatus": "Healthy"
      }
    ]
  }
`
I may be missing something though, so every help is welcome.
Thanks

Deblazz commented:
Hi, 
I converted the dict in an object and changed the OpenApi swagger to reflect the changes in the returned object.
I also added the requested error.
I am open for further feedbacks, 
thanks.

claudiolor commented on frontend/src/app/modules/status/pages/status.component.html:
```diff
@@ -0,0 +1,26 @@
+<div class="mx-auto px-4 py-4 sm:px-8 lg:container 
+            grid grid-cols-1 justify-items-stretch md:grid-cols-3 gap-6 " *transloco="let t">
```
Cleaning up the css classes:
```suggestion
<div class="mx-auto lg:container px-4 py-4 sm:px-8
            grid grid-cols-1 md:grid-cols-3" *transloco="let t">
```

claudiolor commented on backend/pkg/server/handlers/infoClusters.go:
```diff
@@ -0,0 +1,43 @@
+
+	f := factory.Factory{
+		KubeClient: s.nativeClient,
+		CRClient:   s.oClient,
```
We are not passing the Liqo namespace here, which causes the status to be incorrectly determined. The health checker looks for pods in an unhealthy state in the Liqo namespace. However, since the namespace is an empty string, it ends up checking for unhealthy pods across the entire cluster.

I believe we should default to the Liqo namespace, while still allowing the option as a parameter in the server configuration (not forgetting to include this parameter in the values.yaml file of the Helm chart).

claudiolor commented on frontend/src/app/shared/components/peering-table/peering-table.component.ts:
```diff
@@ -0,0 +1,66 @@
+      { headerName: this.translateService.translate('status.peering.roleLabel'), field: 'role' },
+      { headerName: this.translateService.translate('status.peering.networkingStatusLabel'), field: 'networkingStatus' },
+      { headerName: this.translateService.translate('status.peering.authStatusLabel'), field: 'authenticationStatus' },
+      { headerName: this.translateService.translate('status.peering.offloadingStatusLabel'), field: 'offloadingStatus' },
```
If would be nice having colors depending on the status

claudiolor commented on backend/cmd/main.go:
```diff
@@ -39,8 +39,13 @@ func main() {
+
 	// create a type that satisfies the `api.ServerInterface`, which contains an implementation of every operation from the generated code
-	server := handlers.NewServer(oClient)
+	server := handlers.NewServer(oClient, nativeClient)
```
```suggestion
	server := handlers.NewServer(oClient, nativeClient, liqoNamespace)
```

And you can get the namespace as argument as follows:

```
liqoNamespace := flag.String("liqo-namespace", "liqo", "The namespace where liqo is installed")
```

before `flag.Parse()`

claudiolor commented on frontend/src/app/modules/status/pages/status.component.html:
```diff
@@ -0,0 +1,26 @@
+<div class="mx-auto lg:container px-4 py-4 sm:px-8
+            grid grid-cols-1 md:grid-cols-3" *transloco="let t">
```
Add some spacing between when vertically aligned and make sure they cover the entire width
```suggestion
            grid grid-cols-1 md:grid-cols-3 gap-2 w-full" *transloco="let t">
```

claudiolor commented on frontend/src/app/shared/components/peering-table/peering-table.component.ts:
```diff
@@ -0,0 +1,75 @@
+      { headerName: this.translateService.translate('status.peering.roleLabel'), field: 'role' },
+      {
+        headerName: this.translateService.translate('status.peering.networkingStatusLabel'), field: 'networkingStatus', cellClass: params =>
+          params.value?.toLowerCase() === 'healthy' ? 'status-healthy' : 'status-unhealthy',
```
I'd define a function like the following:
```
getStatusClass(status: string): string {
    switch (status.toLowerCase()) {
      case 'healthy':
        return 'status-healthy';
      case 'unhealthy':
        return 'status-unhealthy';
      default:
        return '';
    }
  }
```

Then calling it here.
Additionally, I wouldn't color the text red if the status is unknown, so the function above leave the default black color.

claudiolor reviewed (commented):
The build is currently broken because of a wrong tag, plus some other minor issues

Deblazz commented:
Hi @claudiolor,
sorry for the wrong permissions configured on those files. That was my fault.
I believe the rebase with main is completed.
Open to further feedback, have a nice day.

claudiolor commented:
Merged! Thank you for your contribution! 🎉
